### ISSUE TITLE
Using `Base.summary` instead of `short_show` and better show for grids and fields

### ISSUE THREAD

**Issue by glwagner** [CliMA/Oceananigans.jl#2064]
Right now at the REPL we get

```julia
julia> using Oceananigans

julia> grid = RectilinearGrid(size=3, z=(0, 1), topology=(Flat, Flat, Bounded))
RectilinearGrid{Float64, Flat, Flat, Bounded} on the CPU()
                   domain: x ∈ [1.0, 1.0], y ∈ [1.0, 1.0], z ∈ [0.0, 1.0]
                 topology: (Flat, Flat, Bounded)
        size (Nx, Ny, Nz): (1, 1, 3)
        halo (Hx, Hy, Hz): (0, 0, 1)
grid in x: Flattened
grid in y: Flattened
grid in z: Regular, with spacing 0.3333333333333333
```

so `grid in x:` isn't aligned with the rest of the "data table".

Here's what we get when `z` is stretched:

```julia
julia> grid = RectilinearGrid(size=3, z=[0.0, 0.2, 0.5, 1.0], topology=(Flat, Flat, Bounded))
RectilinearGrid{Float64, Flat, Flat, Bounded} on the CPU()
                   domain: x ∈ [1.0, 1.0], y ∈ [1.0, 1.0], z ∈ [0.0, 1.0]
                 topology: (Flat, Flat, Bounded)
        size (Nx, Ny, Nz): (1, 1, 3)
        halo (Hx, Hy, Hz): (0, 0, 1)
grid in x: Flattened
grid in y: Flattened
grid in z: Stretched, with spacing min=0.2, max=0.5
```

So here's a few ideas:

#### Include "spacing" information under the `domain` row?

One idea would be to show something like

```julia
x: Flat
y: Flat
z: [0.0, 1.0], Δz: 0.33
```

instead of `domain:` (for the regularly-spaced case), and

```julia
x: Flat
y: Flat
z: [0.0, 1.0], extrema(Δz): (0.2, 0.5)
```

for the stretched case. We could also increase the length of the table and put the spacings on their own lines, eg

```julia
x: Flat
y: Flat
z: [0.0, 1.0]
extrema(Δz): (0.2, 0.5)
```

We might also just hide flattened directions (we already express that the directions are `Flat` so why print this again?)

#### Compactify how we show the size + halo. I've always found the way we write `(Nx, Ny, Nz)` to negatively impact readability. 

I think it's ok to just write `size`, because we want to move towards providing a function-based API rather than expecting users to access struct properties directly. Eg we could have just

```julia
size: (1, 1, 1)
halo: (0, 0, 1)
```

What do others think?

**Comment by navidcy** [comment]
- `extrema(Δz)` is not as clear... let's just write `min(Δz) = ..., max(Δz) = ...`
- I'd suggest something like
  ```Julia
  z ∈ [0.0, 1.0], min(Δz) = ..., max(Δz) = ...
  ```

**Comment by tomchor** [comment]
I actually appreciate the verbose-ness of the grid show functions (actually most of Oceananigans show functions in general). So I can only assume that more beginner users appreciate that as well, since it makes extremely clear what the grid characteristics are. I also don't think there's a very important downside to being verbose here, so I'd vote to keep the same style.

Although I agree that we could get rid of some redundant information. My suggestion would be something like:

```julia
julia> grid = RectilinearGrid(size=3, z=[0.0, 0.2, 0.5, 1.0], topology=(Flat, Flat, Bounded))
RectilinearGrid{Float64, Flat, Flat, Bounded} on the CPU()
                   domain: x ∈ [1.0, 1.0], y ∈ [1.0, 1.0], z ∈ [0.0, 1.0]
                 topology: (Flat, Flat, Bounded)
        size (Nx, Ny, Nz): (1, 1, 3)
        halo (Hx, Hy, Hz): (0, 0, 1)
                z spacing: Stretched, with min=0.2, max=0.5
```

In other words, I'd remove the flat dimensions (since that's unnecessary), and clean and align the `z` spacing statement. I'd also keep the ` topology: (Flat, Flat, Bounded)` because the line `RectilinearGrid{Float64, Flat, Flat, Bounded}` might not be very intuitive to everyone.

If a shorter show is necessary in some places maybe we could create something like `short_show`? (Or maybe define `show` differently based on the display?: https://schurkus.com/2018/01/05/julia-print-show-display-dump-what-to-override-what-to-use/)

**Comment by glwagner** [comment]
I like how `julia` treats arrays. The first line is `Base.summary`, which can be re-used for short summaries elsewhere (eg `Base.summary(::RectilinearGrid)` should be used when describing `field.grid`). `Base.summary` provides type and type-like information. For example:

```julia
julia> Base.summary(rand(3, 3))
"3×3 Matrix{Float64}"

julia> Base.summary(OffsetArray(rand(3), -1))
"3-element OffsetArray(::Vector{Float64}, 0:2) with eltype Float64 with indices 0:2"
```

For us type-like information include the size of the grid, the eltype, and the topology, so this might read

```julia
"1×1×3 RectilinearGrid{Float64} on CPU with topology (Flat, Flat, Bounded) and halo (0, 0, 1)"
```

I'm not sure this is a good idea, but we could also omit `Flat` dimensions from the size display, eg

```julia
"3-element RectilinearGrid{Float64} on CPU with topology (Flat, Flat, Bounded) and halo (0, 0, 1)"
```

```julia
"128×128 RectilinearGrid{Float64} on CPU with topology (Periodic, Periodic, Flat) and halo (1, 1, 0)"
```

If we also change `interior` to drop `Flat` dimensions then I think this makes sense (otherwise it might not).

In addition to type information, `show` also gives information about _values_. Here this is the extent of the domain and the grid spacings. For example:

```julia
julia> rand(3, 3)
3×3 Matrix{Float64}:
 0.460633  0.134952   0.0366328
 0.6353    0.0632607  0.0657011
 0.339457  0.720342   0.0417144
```

for this we can use a tabular display like `domain: x ∈ [1.0, 1.0], y ∈ [1.0, 1.0], z ∈ [0.0, 1.0]`. However, I think a better possibility is to explicitly show the locations of the faces using a strategy similar to how `julia` shows row vectors:

```julia
julia> collect(reshape(0:0.1:10, 1, 101))
1×101 Matrix{Float64}:
 0.0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9  1.0  1.1  1.2  …  8.9  9.0  9.1  9.2  9.3  9.4  9.5  9.6  9.7  9.8  9.9  10.0
```

so we might have, depending on the width of the display:

```julia
1×1×101 RectilinearGrid{Float64} on CPU with topology (Flat, Flat, Bounded) and halo (0, 0, 1)
z faces: [0.0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  0.8  0.9   …  9.2  9.3  9.4  9.5  9.6  9.7  9.8  9.9  10.0]
```

etc. Or, putting a "domain summary" out front, something like

```julia
1×1×101 RectilinearGrid{Float64} on CPU with topology (Flat, Flat, Bounded) and halo (0, 0, 1)
z ∈ [0.0, 10.0] with faces [0.0  0.1  0.2  0.3  0.4  0.5  0.6  0.7  …  9.4  9.5  9.6  9.7  9.8  9.9  10.0]
```

We can do something similar for spacings, though as others have noted I think it is useful to show _statistics_ rather than list all the values for that case. We probably want to indicate whether we are talking about spacings at Center or Face, so

```julia
z spacings: stretched with minimum(Δzᵃᵃᶜ) = 0.2, maximum(Δzᵃᵃᶜ) = 0.5
```

is an option.

**Comment by navidcy** [comment]
I'm actually happy with how `show` looks atm.

E.g.,

```julia
julia> grid = RectilinearGrid(size = (Nx, Ny, Nz),
                              topology=(Periodic, Bounded, Bounded),
                              x = (0, Lx),
                              y = chebychev_spaced_y_faces,
                              z = chebychev_spaced_z_faces)
RectilinearGrid{Float64, Periodic, Bounded, Bounded} on the CPU()
                    domain: x ∈ [0.0, 10000.0], y ∈ [-5000.0, 5000.0], z ∈ [-1000.0, 0.0]
                  topology: (Periodic, Bounded, Bounded)
         size (Nx, Ny, Nz): (64, 64, 32)
         halo (Hx, Hy, Hz): (1, 1, 1)
                 grid in x: Regular, with spacing 156.25
                 grid in y: Stretched, with spacing min=6.022718974138115, max=245.33837163709035
                 grid in z: Stretched, with spacing min=2.407636663901485, max=49.008570164780394
```

If anything, I'd suggest the last three lines to be changed to:


```julia
julia> grid = RectilinearGrid(size = (Nx, Ny, Nz),
                              topology=(Periodic, Bounded, Bounded),
                              x = (0, Lx),
                              y = chebychev_spaced_y_faces,
                              z = chebychev_spaced_z_faces)
RectilinearGrid{Float64, Periodic, Bounded, Bounded} on the CPU()
                    domain: x ∈ [0.0, 10000.0], y ∈ [-5000.0, 5000.0], z ∈ [-1000.0, 0.0]
                  topology: (Periodic, Bounded, Bounded)
         size (Nx, Ny, Nz): (64, 64, 32)
         halo (Hx, Hy, Hz): (1, 1, 1)
              spacing in x: Regular, with spacing 156.25
              spacing in y: Stretched, with spacing min=6.022718974138115, max=245.33837163709035
              spacing in z: Stretched, with spacing min=2.407636663901485, max=49.008570164780394
```

**Comment by glwagner** [comment]
Agree the alignment is fixed so this isn't too urgent, and that what we currently do is adequate.

I would still like to work on a better `Base.summary`, which will change some of `show` (mostly how we display the grid size, architecture, and topology). But we can keep the parts we like, such as how we show the coordinate nodes and spacings.
